### PR TITLE
Fix bugs in parent class init/cleanup logic

### PR DESCRIPTION
**Pull request by NGloreous**
Fixing two bugs related to feature #143 
- Call base class cleanup even if top level test class doesn't have class cleanup
- Fix null ref when a class more than 1 level of inheritance has a class init/cleanup method.

### CODE REVIEW

**Comment by NGloreous** [comment]
@parrainc FYI since this fixes some bugs from your change in case you have feedback.

File: src/Adapter/MSTest.CoreAdapter/Execution/TestClassInfo.cs [hunk 1]
```diff
@@ -355,16 +365,16 @@ public void RunClassInitialize(TestContext testContext)
         [SuppressMessage("Microsoft.Design", "CA1031:DoNotCatchGeneralExceptionTypes", Justification = "Requirement is to handle all kinds of user exceptions and message appropriately.")]
         public string RunClassCleanup()
         {
-            if (this.ClassCleanupMethod is null && !this.BaseClassInitAndCleanupMethods.Any(p => p.Item2 != null))
+            if (this.ClassCleanupMethod is null && !this.BaseClassCleanupMethodsStack.Any())
```
Comment (jayaranigarg) [inline src/Adapter/MSTest.CoreAdapter/Execution/TestClassInfo.cs]: We should keep this check as it is for the cases where we don't have classInitalize but we do have classCleanUp methods.

We are checking for this.BaseClassCleanupMethodsStack.Any() in the below if check as well.

File: src/Adapter/MSTest.CoreAdapter/Execution/TypeCache.cs [hunk 2]
```diff
@@ -485,17 +485,15 @@ private bool IsAssemblyOrClassCleanupMethod(MethodInfo methodInfo, Type cleanupA
             TestClassInfo classInfo,
             ref MethodInfo[] initAndCleanupMethods)
         {
-            if (initAndCleanupMethods is null)
+            if (initAndCleanupMethods.Any(x => x != null))
```
Comment (jayaranigarg) [inline src/Adapter/MSTest.CoreAdapter/Execution/TypeCache.cs:488]: What is the motivation for this change? Why do we need to reinitialize initAndCleanupMethods?

File: test/UnitTests/MSTest.CoreAdapter.Unit.Tests/Execution/TestClassInfoTests.cs [hunk 3]
```diff
@@ -500,6 +500,20 @@ public void RunClassCleanupShouldReturnExceptionDetailsOfNonAssertExceptions()
                 "Class Cleanup method DummyTestClass.ClassCleanupMethod failed. Error Message: System.ArgumentException: Argument Exception. Stack Trace:     at Microsoft.VisualStudio.TestPlatform.MSTestAdapter.UnitTests.Execution.TestClassInfoTests.<>c.<RunClassCleanupShouldReturnExceptionDetailsOfNonAssertExceptions>");
         }
 
+        [TestMethod]
+        public void RunBaseClassCleanupEvenIfThereIsNoTopLevelClassCleanup()
```
Comment (jayaranigarg) [inline test/UnitTests/MSTest.CoreAdapter.Unit.Tests/Execution/TestClassInfoTests.cs]: Should we rename this to 'RunBaseClassCleanupEvenIfThereIsNoDerivedClassCleanup' ? 
TopLevel seems a bit unclear.

File: src/Adapter/MSTest.CoreAdapter/Execution/TypeCache.cs (hunk 2 above)
Comment (MatthewSteeples) [reply src/Adapter/MSTest.CoreAdapter/Execution/TypeCache.cs:488]: If you run the included tests but without the fixes you can see the problems that are being solved. Because this method used to reset `initAndCleanupMethods` back to null, it would cause a NullReferenceException elsewhere in the loop. So rather than the array being null, just the elements are null in this new process

**Comment by MatthewSteeples** [comment]
We're seeing some problems with this change (we needed the same fix, so we've got a build of it working). I believe (not 100% sure) that it's changed the order of execution for TestInitialize attributes

**Comment by MatthewSteeples** [comment]
Turns out that the scenario that these changes had enabled for us exposed a flaw in our tests that had runaway memory usage, so the problem was in our code! I think it just exhausted the memory on the Pipeline servers until paging stuff in and out to disk caused them to be unresponsive. When we got it working on a locally installed build agent it topped out at 24gb of RAM in use!

File: test/UnitTests/MSTest.CoreAdapter.Unit.Tests/Execution/TestClassInfoTests.cs (hunk 3 above)
Comment (NGloreous) [reply test/UnitTests/MSTest.CoreAdapter.Unit.Tests/Execution/TestClassInfoTests.cs]: Sure

File: src/Adapter/MSTest.CoreAdapter/Execution/TestClassInfo.cs (hunk 1 above)
Comment (NGloreous) [reply src/Adapter/MSTest.CoreAdapter/Execution/TestClassInfo.cs]: I can revert it back but it's basically an equivalent check because things don't get pushed into BaseClassCleanupMethodsStack unless Item2 in BaseClassInitAndCleanupMethods was not null.

File: src/Adapter/MSTest.CoreAdapter/Execution/TypeCache.cs (hunk 2 above)
Comment (NGloreous) [reply src/Adapter/MSTest.CoreAdapter/Execution/TypeCache.cs:488]: Yes what Matt says is correct. This is the main fix for the null ref issue.

**Comment by parrainc** [comment]
> @parrainc FYI since this fixes some bugs from your change in case you have feedback.

I think already provided review comments covers pretty much it. For me, it looks good! Thanks for addressing that issue :)
